UI/TechTreeWnd.cpp

```diff
@@ -1583,20 +1604,108 @@ void TechTreeWnd::TechListBox::TechRow::Render() {
 }
 
 std::vector<GG::X> TechTreeWnd::TechListBox::TechRow::ColWidths(GG::X total_width) {
-    const GG::X GRAPHIC_WIDTH(ClientUI::Pts() * 2);
-    const GG::X NAME_WIDTH(ClientUI::Pts() * 18);
-    const GG::X COST_WIDTH(ClientUI::Pts() * 4);
-    const GG::X TIME_WIDTH(ClientUI::Pts() * 4);
-    const GG::X CATEGORY_WIDTH(ClientUI::Pts() * 8);
-
-    const GG::X DESC_WIDTH = std::max(GG::X1, total_width - GRAPHIC_WIDTH - NAME_WIDTH - COST_WIDTH - TIME_WIDTH - CATEGORY_WIDTH);
     std::vector<GG::X> retval;
-    retval.push_back(GRAPHIC_WIDTH);
-    retval.push_back(NAME_WIDTH);
-    retval.push_back(COST_WIDTH);
-    retval.push_back(TIME_WIDTH);
-    retval.push_back(CATEGORY_WIDTH);
-    retval.push_back(DESC_WIDTH);
+    // Early return for very small screen
+    if (total_width < (6 * 4 * ClientUI::Pts())) {
+        ErrorLogger() << "Not enough room for reasonable column space";
+        retval = {GG::X1, GG::X1, GG::X1, GG::X1, GG::X1, GG::X1};
+        return retval;
+    }
+
+    // Attempt to utilize user defined widths
+    // If not enough width to fit, attempt default widths
+    // Failing both cases, proportionally assign the available width
+
+    GG::X graphic_width(    GetOptionsDB().Get<int>(TECHROW_COL_XML_PREFIX + ".graphic"));
+    GG::X name_width(       GetOptionsDB().Get<int>(TECHROW_COL_XML_PREFIX + ".name"));
+    GG::X cost_width(       GetOptionsDB().Get<int>(TECHROW_COL_XML_PREFIX + ".cost"));
+    GG::X time_width(       GetOptionsDB().Get<int>(TECHROW_COL_XML_PREFIX + ".time"));
+    GG::X category_width(   GetOptionsDB().Get<int>(TECHROW_COL_XML_PREFIX + ".category"));
+
+    GG::X cols_width_sum = graphic_width + name_width + cost_width + time_width + category_width;
+
+    GG::X desc_width(std::max(GG::X1, total_width - cols_width_sum));
+
+    if (cols_width_sum > total_width) {
+        // config values did not fit, attempt default values
+        graphic_width   = GG::X(GetOptionsDB().GetDefault<int>(TECHROW_COL_XML_PREFIX + ".graphic"));
```

Comment:
all this prefix stuff isn't really saving enough space to warrant hiding the actual value amongst several variables / constants, I think...